File: Data Structures and Algorithms/Algorithms/javascript/binarySearchTree.js
class Node{
    constructor(value){
        this.value = value;
        this.left = null;
        this.right = null;
    }
}
class BinarySearchTree{
    constructor(){
        this.root = null;
    }
    insert(value){
        var newNode = new Node(value);
        if(this.root === null){
        this.root = newNode;
        return this;
        }
        var current = this.root;
        while(true){
            if(value === current.value) return undefined;
            if(value < current.value){
                if(current.left === null){
                    current.left = newNode;
                    return this;
                }
                current = current.left
            }else{
                if(current.right === null){
                    current.right = newNode;
                return this;
                }
                current = current.right;
                    }
                    }
            }
    find(value){
        var matchNode = new Node(value);
        if(!this.root) return undefined;
        var currentVal = this.root;
        var match = false;
        while(!match && currentVal){
            if(value < currentVal.value){
                currentVal = currentVal.left;
            }else if(value > currentVal.value){
                currentVal = currentVal.right;
                }else{
                    return true;
                }
        }
        return false;
    }
    breadFirstSearch(){
        var data = [];
        var queue = [];
        var node = this.root;
        queue.push(node);
        while(queue.length){
            node = queue.shift();
            data.push(node.value);
            if(node.left) queue.push(node.left);
            if(node.right) queue.push(node.right);}
                return data; }
    depthFirstSearchPREORDER(){
        var data = [];
        var current = this.root;
        function traverse(node){
            data.push(node.value);
            if(node.left) traverse(node.left);
            if(node.right) traverse(node.right);}
        traverse(current);
        return data;
    }
    depthFirstSearchPOSTORDER(){
        var data = [];
        var current = this.root;
        function traverse(node){
            if(node.left) traverse(node.left);
            if(node.right) traverse(node.right);
            data.push(node.value);}
            traverse(this.root);
            return data;
        }
}

File: DataStructuresandAlgorithms/DataStructures/BinarySearchTree.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *left;
    struct node *right;
};

struct node *newNode(int item)
{
    struct node *t = (struct node *)malloc(sizeof(struct node));
    t->data=item;
    t->left=t->right=NULL;
}

struct node* insert(struct node* node, int data )
{
    if(node==NULL)
    {
        return newNode(data);
    }

    if(data < node->data)
    {
        node->left=insert(node->left,data);
    }
    else if(data> node->data)
    {
        node->right=insert(node->right,data);
    }

    return node;
}

void inorder(struct node *root)
{
    if(root!=NULL)
    {
        inorder(root->left);
        printf("%d ",root->data);
        inorder(root->right);
    }
}

/*void preorder(struct node *root)
{
    if(root!=NULL)
    {
        printf("%d ",root->data);
        inorder(root->left);
        inorder(root->right);
    }
}*/

void postorder(struct node *root)
{
    if(root!=NULL)
    {
        postorder(root->left);
        postorder(root->right);
        printf("%d ",root->data);
    }
}



int main()
{
    struct node *root = NULL;
    root=insert(root,83);
    insert(root,25);
    insert(root,11);
    insert(root,12);
    insert(root,32);
     insert(root,95);
      insert(root,46);
       insert(root,87);
        insert(root,68);
         insert(root,100);
    
    //inorder(root);
    printf("\n");
    //preorder(root);
    printf("\n");
    postorder(root);

}